equinox/nn/_attention.py

```diff
@@ -308,6 +391,58 @@ def __call__(
                     "process_heads must not change the shape of the heads."
                 )
 
+        if state is None:
+            causal_mask_offset = 0
+        else:
+            if self.autoregressive_index is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., state_length=...)`."
+                )
+            key_state, value_state, index = state.get(self.autoregressive_index)
+            key_seq_length = key_heads.shape[0]
+            if index + key_seq_length > self.state_length:
```

Comment:
These `if` statements will fail under JIT.

FWIW I suspect it won't be possible to make things work efficiently at the end anyway, and it's not obvious to me that rolling is really the better choice. (E.g. some techniques make sure to use the start of an attention window as a place to put globally-attended-to information.) I think I'm probably happy with the behaviour without this.

Reply:
I see. I reverted it back to the previous implementation.

AFAICT in the current implementation of the cache, if the cache is:

`1 2 3 4 5 6 7 0 0`

and the `seq_len` is 3, then after the update it becomes:

`1 2 3 4 5 6 A B C` 

But if I understood you correctly (from your previous comment), you want it to be 

`1 2 3 4 5 6 7 A B` 

right? In the rolling strategy, it would have become:

`2 3 4 5 6 7 A B C`.

 Perhaps David could weigh in on this, because I don't know what the best approach is. @dlwh WDYT?

Reply:
A very neat explanation! Yes, that's it exactly.

Reply:
I've stayed away from inference so far, so I mostly haven't had to think about caching. That seems sensible and clean.

It does seem like a thing this could be pluggable? I know it's built on State right now, but if the KVCache were a module with an abstract method, you could choose the behavior.

The MaxText code is pretty hard to follow (for me) but I think there are two caches there, one that does the prefilling/prefix that Patrick talks about, and another that's for the ar decoding. [this](https://github.com/google/maxtext/blob/1ae752103af6f2939fbb3f6bbbfa998eee8e4d76/MaxText/layers/attentions.py#L653) means it's doing a roll for AR tokens and leaving the prefill tokens in place. I think?

Reply:
> It does seem like a thing this could be pluggable? I know it's built on State right now, but if the KVCache were a module with an abstract method, you could choose the behavior.

I really like this idea! I can cook something up. It would also make the already pretty large MHA file slightly more svelte.